handlers/product-move-api/src/main/scala/com/gu/productmove/refund/RefundSupporterPlus.scala

```diff
@@ -117,24 +102,49 @@ object RefundSupporterPlus {
     // - one for the charge where the SourceType is "InvoiceDetail" and SourceId is the invoice item id
     // - one for the tax where the SourceType is "Tax" and SourceId is the taxation item id
     // https://www.zuora.com/developer/api-references/older-api/operation/Object_POSTInvoiceItemAdjustment/#!path=SourceType&t=request
-    invoiceItems.filter(_.amountWithTax != 0).flatMap { invoiceItem =>
+
+    // Charges will have a negative value because we are dealing with the negative invoice created by cancellation,
+    // discounts will have a positive value. Unfortunately we can't just create an adjustment for each of the charges
+    // and discounts as this takes the invoice to values which are not allowed by Zuora. Instead we need to subtract
+    // the discount from the charge and create an adjustment for the difference.
+    val (charges, discounts) = invoiceItems
+      .filter(_.amountWithTax != 0) // Ignore any items with a zero amount, they will be the contribution charge where the user is just paying the base price
+      .partition(_.amountWithTax < 0)
+    val maybeDiscount = discounts.headOption
+    val discountedCharge = maybeDiscount.flatMap(discount =>
+      charges.find(charge =>
+        charge.ChargeAmount.abs >= discount.ChargeAmount &&
+          charge.taxAmount.abs >= discount.taxAmount,
+      ),
+    )
```

Comment:
In the normal case (especially for a newish and discounted S+) we expect there to be only 1 charge and one discount.
so we assume there's only one discount is there at the most, and then we find (any/the) charge that is bigger than the discount and subtract it from that.
Is that all about right?

Reply:
we are using a query, and we are only getting Id, ChargeAmount, TaxAmount, ChargeDate, InvoiceId

https://github.com/guardian/support-service-lambdas/blob/45bd77421a85caa15c234efcec371b3e3d0b8c57/handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/GetRefundInvoiceDetails.scala#L30

however you could add an extra couple of fields maybe `RatePlanChargeId`  would match between the two, or failing that `AppliedToChargeNumber`  and `ChargeNumber` to correlate the discount and charge, see
https://knowledgecenter.zuora.com/Zuora_Central_Platform/API/G_SOAP_API/E1_SOAP_API_Object_Reference/InvoiceItem

Reply:
this flatmap and the one below would probably be neater as for comprehensions
```suggestion
    val discountedCharge = for {
      discount <- maybeDiscount
    } yield charges.find(charge =>
      charge.ChargeAmount.abs >= discount.ChargeAmount &&
        charge.taxAmount.abs >= discount.taxAmount,
    )
```

Reply:
There will be two charges, one for the subscription and one for the contribution but the contribution one will usually be zero. Then there may also be at most one discount, in which case we find any charge where the charge amount AND the tax amount are larger than the discount charge amount and tax amount.